sklearn/utils/estimator_checks.py

```diff
@@ -2046,9 +2048,21 @@ def check_estimators_data_not_an_array(name, estimator_orig, X, y):
     estimator_2 = clone(estimator_orig)
     set_random_state(estimator_1)
     set_random_state(estimator_2)
-
-    y_ = NotAnArray(np.asarray(y))
-    X_ = NotAnArray(np.asarray(X))
+    
+    if obj_type not in ["NotAnArray", 'PandasDataframe']:
+        raise ValueError("Data type {0} not supported".format(obj_type))
+        
+    if obj_type == "NotAnArray":
+        y_ = NotAnArray(np.asarray(y))
+        X_ = NotAnArray(np.asarray(X))
+    else:
+        import pandas as pd
```

Comment:
We don't always have `pandas` in our test environments. Pandas is not officially a dependency and therefore we make sure that the code base works without it. As a result all tests which use `pandas` should use it with `pd = pytest.importorskip('pandas')` instead. `pytest` will simply skip the test if pandas is not installed.

Reply:
A search of the sklearn folder showed this syntax for importing pandas, but all the files were "test_*.py" files. Is it correct to add this syntax to the file in this PR, or should it be going somewhere else?

Reply:
That's a really good point @reshamas. It may be a good idea to catch the `ImportError` here and if `obj_type` is `PandasDataframe` and import fails, to give a `ValueError` or something, since that basically means the input is not compatible with the environment. Then when the tests want to call this function, they should skip the tests that would give this error using `pytest.importorskip`, I guess.

I'm really not sure which exception is more apt to raise though!

Reply:
Would it be something like this:   
```python
import pytest
from pytest import importorskip
```

```python
    if obj_type not in ["NotAnArray", 'PandasDataframe']:
        raise ValueError("Data type {0} not supported".format(obj_type))

    if obj_type == "NotAnArray":
        y_ = NotAnArray(np.asarray(y))
        X_ = NotAnArray(np.asarray(X))
    elif obj_type == "PandasDataframe":
        try:  
            import pandas as pd
            pd = pytest.importorskip("pandas")
            y_ = np.asarray(y)
            if y_.ndim == 1:
                y_ = pd.Series(y_)
            else:
                y_ = pd.DataFrame(y_)
            X_ = pd.DataFrame(np.asarray(X))
        except ImportError:
            raise SkipTest("Skipping test, pandas not installed")
```
OR
```python
except ImportError:
    assert_raise_message(ValueError, "Skipping test, pandas not installed ")
--

Reply:
More like the latter option, but maybe say "Cannot run a pandas related test when pandas is not installed." or something. Cause we're not really _skipping_ here, were' raising an error.

Reply:
@reshamas @adrinjalali Thanks for your suggestions. I will make the changes accordingly.